### PR TITLE
feat: Add test cases for PATCH /requests/:id API for acknowledging OOO requests

### PR DESCRIPTION
**Pull request by surajmaity1**
Date: 22 Mar, 2025

Developer Name:  Suraj Maity ( @surajmaity1 )

---

## Issue Ticket Number

- https://github.com/Real-Dev-Squad/website-backend/issues/2242
## Description

This PR adds test cases for the PATCH /requests/:id API for acknowledging OOO requests to ensure its correctness and reliability. The tests cover the following scenarios:
- Valid input for acknowledging OOO request.
- Authentication and authorization checks.
- Error handling for edge cases and unexpected failures.
### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Test Coverage
<details>
<summary>Screenshot 1</summary>

integration

image
image
midd

image

acknowledge test Coverage summary

</details>

## Additional Notes

- Feature PR https://github.com/Real-Dev-Squad/website-backend/pull/2384

### CODE REVIEW

**Comment by prakashchoudhary07** [comment]
@surajmaity1 Error in build it self, please have a look

**Comment by prakashchoudhary07** [comment]
@surajmaity1 Build is faling, please check

**Comment by surajmaity1** [comment]
> @surajmaity1 Build is faling, please check

Hi @prakashchoudhary07 , it's fixed.
Thank you.

**Comment by surajmaity1** [comment]
> @surajmaity1 Error in build it self, please have a look

Hi @prakashchoudhary07 , it's fixed now.
Thank you.

File: test/integration/requests.test.ts [hunk 1]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
     });
   });
 
+  describe.skip("PATCH /requests/:id", function () {
+    let pendingOooRequestId1: string;
+    let oooRequestData3: any;
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:328]: why type any?

File: test/integration/requests.test.ts [hunk 2]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+    let invalidRequestId: string;
+    let approvedOooRequestId: string;
+    let rejectedOooRequestId: string;
+
+    beforeEach(async function () {
+      oooRequestData3 = { ...createOooRequests3, userId: testUserId };
+
+      const { id: pendingOooId1 }: any = await createRequest(oooRequestData3);
+      pendingOooRequestId1 = pendingOooId1;
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:337]: You could simplify this by directly destructuring the id into pendingOooRequestId1

File: test/integration/requests.test.ts [hunk 3]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+
+      const { id: pendingOooId2 }: any = await createRequest(oooRequestData3);
+
+      const { id: pendingOooId3 }: any = await createRequest(oooRequestData3);
+
+      const { id: invalidId }: any = await createRequest({...oooRequestData3, type: "XYZ"});
+      invalidRequestId = invalidId;
+
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:345]: same here

File: test/integration/requests.test.ts [hunk 4]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+      const { id: approvedOooId }: any = await updateRequest(pendingOooId2, { status: REQUEST_STATE.APPROVED }, testSuperUserId, REQUEST_TYPE.OOO);
+      approvedOooRequestId = approvedOooId;
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:347]: why are we passing type here wherever above request don't need it?

File: test/integration/requests.test.ts [hunk 5]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+
+      const { id: rejectedOooId }: any = await updateRequest(pendingOooId3, { status: REQUEST_STATE.REJECTED }, testSuperUserId, REQUEST_TYPE.OOO);
```
Comment (MayankBansal12) [inline test/integration/requests.test.ts:349]: please avoid using any

File: test/integration/requests.test.ts [hunk 6]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+      rejectedOooRequestId = rejectedOooId;
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:350]: destructure here also

File: test/integration/requests.test.ts [hunk 7]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+        .request(app)
+        .patch(`/requests/${pendingOooRequestId1}?dev=true`)
+        .set("cookie", `${cookieName}=${authToken}`)
+        .send(acknowledgeOooRequest)
+        .end(function (err, res) {
+          if (err) {
+            return done(err);
+          }
+          expect(res.statusCode).to.equal(403);
+          // expect(res.body.message).to.equal(UNAUTHORIZED_TO_ACKNOWLEDGE_OOO_REQUEST);
```
Comment (MayankBansal12) [inline test/integration/requests.test.ts:409]: please remove this comment

File: test/integration/requests.test.ts [hunk 8]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
     });
   });
 
+  describe.skip("PATCH /requests/:id", function () {
+    let pendingOooRequestId1: string;
+    let oooRequestData3: any;
+    let invalidRequestId: string;
+    let approvedOooRequestId: string;
+    let rejectedOooRequestId: string;
```
Comment (MayankBansal12) [inline test/integration/requests.test.ts:331]: please fix the naming

File: test/integration/requests.test.ts [hunk 9]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+          expect(res.statusCode).to.equal(501);
+          expect(res.body.message).to.equal("Feature not implemented");
+          done();
+        });
+    });
+
+    it("should return 404 if request does not exist", function (done) {
+      chai
+        .request(app)
+        .patch(`/requests/111111?dev=true`)
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:385]: why don't we use invalidRequestId here?

**Review by AnujChhikara — CHANGES_REQUESTED** [review]

File: test/integration/requests.test.ts (hunk 7 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts:409]: I'm importing UNAUTHORIZED_TO_ACKNOWLEDGE_OOO_REQUEST from feature PR.
So, if I remove this comment, it'll show - `Cannot find name 'UNAUTHORIZED_TO_ACKNOWLEDGE_OOO_REQUEST'`

File: test/integration/requests.test.ts [hunk 10]
```diff
@@ -316,6 +323,191 @@ describe("/requests OOO", function () {
...
+            return done(err);
+          }
+          expect(res.statusCode).to.equal(409);
+          expect(res.body.message).to.equal(REQUEST_ALREADY_REJECTED);
+          done();
+        });
+    });
+
+    it("should return 400 if request type is not OOO", function (done) {
+      chai
```
[the remark this answers is not in the document]
Comment (surajmaity1) [reply test/integration/requests.test.ts:445]: Hi  @AnujChhikara  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 9 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @AnujChhikara  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 6 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @AnujChhikara  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 4 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @AnujChhikara  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 3 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts:343]: Hi @AnujChhikara  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 2 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @AnujChhikara ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 1 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @AnujChhikara  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 8 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @MayankBansal12  ,
it's fixed.

Thank you!

File: test/integration/requests.test.ts (hunk 5 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts]: Hi @MayankBansal12 ,
it's fixed.

Thank you!

File: test/fixtures/oooRequest/oooRequest.ts [hunk 11]
```diff
@@ -157,4 +157,19 @@ export const updateOooStatusRequest = [
...
+export const createOooRequests3 = {
+  from: Date.now() + 100000,
+  until: Date.now() + 200000,
+  type: "OOO",
+  requestedBy: "suraj-maity-1",
+  reason: "Out of office for personal emergency.",
+  status: REQUEST_STATE.PENDING
+};
+
+export const acknowledgeOooRequest = {
```
Comment (AnujChhikara) [inline test/fixtures/oooRequest/oooRequest.ts:171]: Should we rename this to acknowledgedOooRequest?

File: test/integration/requests.test.ts [hunk 12]
```diff
@@ -316,6 +323,189 @@ describe("/requests OOO", function () {
...
+  describe.skip("PATCH /requests/:id", function () {
+    let testOooRequest;
+    let onboardingRequest;
+    let approvedOooRequest;
+    let rejectedOooRequest;
+
+    beforeEach(async function () {
+
+      oooRequestData = { ...createOooRequests3, userId: testUserId };
+      testOooRequest = await createRequest(oooRequestData);
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:335]: do we need the whole request data if not then we can just do this 
```suggestion
      testOooRequestId = await createRequest(oooRequestData).id;
```

File: test/integration/requests.test.ts [hunk 13]
```diff
@@ -316,6 +323,189 @@ describe("/requests OOO", function () {
...
+
+      onboardingRequest = await createRequest({
+        type: REQUEST_TYPE.ONBOARDING,
+        numberOfDays: 5,
+        reason: "This is the reason",
+        userId: testUserId,
+      });
```
Comment (AnujChhikara) [inline test/integration/requests.test.ts:342]: can we please store this data in fixtures or use already present data?

File: test/integration/requests.test.ts (hunk 12 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts:335]: We can't do like this.
Thanks

File: test/integration/requests.test.ts (hunk 13 above)
Comment (surajmaity1) [reply test/integration/requests.test.ts:342]: As we don't have any fixture for request and onboarding, so I kept it here. In future I'll clean this one after onboarding or request fixture.

**Review by Achintya-Chatterjee — APPROVED** [review]

File: test/integration/requests.test.ts (hunk 12 above)
Comment (AnujChhikara) [reply test/integration/requests.test.ts:335]: understood

**Review by AnujChhikara — APPROVED** [review]
